### PR TITLE
8319889: Vector API tests trigger VM crashes with -XX:+StressIncrementalInlining

### PR DESCRIPTION
**Pull request by jatin-bhateja**
This bug fix patch fixes a crash occurring due to combined effect of bi-morphic inlining, exception handling, randomized incremental inlining. In this case top level slice API is invoked using concrete 256 bit vector, some of the intermediate APIs within sliceTemplate are marked for lazy inlining due to randomized IncrementalInlining, these APIs returns an abstract vector which when used for virtual dispatch of subsequent APIs results into bi-morphic inlining on account of multiple profile based receiver types. Consider following code snippet.

```
    ByteVector sliceTemplate(int origin, Vector<Byte> v1) {
        ByteVector that = (ByteVector) v1;
        that.check(this);
        Objects.checkIndex(origin, length() + 1);
        VectorShuffle<Byte> iota = iotaShuffle();
        VectorMask<Byte> blendMask = iota.toVector().compare(VectorOperators.LT, (broadcast((byte)(length() - origin))));  [A]
        iota = iotaShuffle(origin, 1, true);     [B]    
        return that.rearrange(iota).blend(this.rearrange(iota), blendMask); [C]
    }
```

Receiver for sliceTemplate is a 256 bit vector, parser defers inlining of toVector() API (see code at line A) and generates a Call IR returning an abstract vector. This abstract vector then virtually dispatches compare API. Compiler observes multiple profile based receiver types (128 and 256 bit byte vectors) for compare API and parser generates a chain of PredictedCallGenerators for bi-morphically inlining it.
      
         PredictedCallGenerators (Vector.compare)
              PredictedCallGenerators (Byte256Vector.compare)
                 ParseGenerator (Byte256Vector.compare)               [D]
                 UncommonTrap (receiver other than Byte256Vector)
              PredictedCallGenerators (Byte128Vector.compare)
                 ParseGenerator (Byte128Vector.compare)               [E]
                 UncommonTrap (receiver other than Byte128Vector)     [F]
              PredictedCallGenerators (UncommonTrap)
         [converged state] = Merge JVM State orginating from C and E  [G]

   Since top level receiver of sliceTemplate is Byte256Vector hence while executing the call generator for Byte128Vector.compare (see code at line E) compiler observes a mismatch b/w incoming argument species i.e. one argument is a 256 bit vector while other is 128 bit vector and throws an exception.
    
At state convergence point (see code at line G), since one of the control path resulted into an exception, compiler propagates the JVM state of other control path comprising of Byte256Mask to downstream graph after  bookkeeping the pending exception state.

Similar to toVector API, iotaShuffle (see code at line B) is also lazily inlined and returns an abstract vector which results into bi-morphic inlining of rearrange. State convergence due to bi-morphic inlining of rearrange results into generation of an abstract ByteVector (Phi Byte128Vector Byte256Vector) which further causes bi-morphic inlining of blend API due to multiple profile based receiver types.
    
Byte128Vector.blend [Java implementation](https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java#L412) explicitly cast incoming mask (Byte256Mask) by Byte128Mask type which results into a[ null reference](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/graphKit.cpp#L1417), this causes a crash while unboxing the mask during inline expansion of blend. To be safe, relaxing the null checking constraint during unboxing to disable intrinsification.

All existing Vector API JTREG tests are passing with -XX:+StressIncrementalInlining at various AVX levels.

Please review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319889](https://bugs.openjdk.org/browse/JDK-8319889): Vector API tests trigger VM crashes with -XX:+StressIncrementalInlining (**Bug** - P4)

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18282/head:pull/18282` \
`$ git checkout pull/18282`

Update a local copy of the PR: \
`$ git checkout pull/18282` \
`$ git pull https://git.openjdk.org/jdk.git pull/18282/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18282`

View PR using the GUI difftool: \
`$ git pr show -t 18282`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18282.diff">https://git.openjdk.org/jdk/pull/18282.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18282#issuecomment-1995081719)

### CODE REVIEW

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 1]
```diff
@@ -161,13 +161,13 @@ Node* GraphKit::box_vector(Node* vector, const TypeInstPtr* vbox_type, BasicType
 
 Node* GraphKit::unbox_vector(Node* v, const TypeInstPtr* vbox_type, BasicType elem_bt, int num_elem, bool shuffle_to_vector) {
   assert(EnableVectorSupport, "");
-  const TypeInstPtr* vbox_type_v = gvn().type(v)->is_instptr();
-  if (vbox_type->instance_klass() != vbox_type_v->instance_klass()) {
-    return nullptr; // arguments don't agree on vector shapes
-  }
-  if (vbox_type_v->maybe_null()) {
+  const TypePtr* vbox_type_v = gvn().type(v)->isa_ptr();
```
Comment (iwanowww) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:164]: You can use `isa_instptr()` and check for `nullptr` instead.

```
  const TypeInstPtr* vbox_type_v = gvn().type(v)->isa_instptr();
  if (vbox_type_v == nullptr || vbox_type->instance_klass() != vbox_type_v->instance_klass()) {
    return nullptr; // arguments don't agree on vector shapes
  }
```

**Review by iwanowww — APPROVED** [review]
Overall, both fixes look good. 

I suggest to handle the bugs separately (as 2 bug fixes).

**Comment by jatin-bhateja** [comment]
> Overall, both fixes look good.
> 
> I suggest to handle the bugs separately (as 2 bug fixes).

Removed shift pattern related changes and created a separate JBS entry for it https://bugs.openjdk.org/browse/JDK-8328309

**Review by iwanowww — APPROVED** [review]

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by TobiHartmann** [comment]
I'm late here but is there a reason no regression test was added?

**Pull request closed without merging**